src/main/java/org/candlepin/resource/util/EntitlementEnvironmentFilter.java

```diff
@@ -0,0 +1,207 @@
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * The EntitlementEnvironmentFilter class filters the consumer's entitlement
+ * whose entitlement certificates are needed to be re-generated because of changes
+ * (addition, removal or re-prioritization) in consumer existing environment.
+ * This filtering avoids re-generating all the consumer entitlement certificates.
+ */
+public class EntitlementEnvironmentFilter {
+    private EntitlementCurator entitlementCurator;
+    private EnvironmentContentCurator environmentContentCurator;
+    private List<String> preExistingEnvs;
+    private List<String> updatedEnvs;
+    private List<String> consumerIds;
+    private List<String> envRemoved;
+
+
+    public EntitlementEnvironmentFilter(EntitlementCurator entitlementCurator,
+        EnvironmentContentCurator environmentContentCurator) {
+        this.entitlementCurator = entitlementCurator;
+        this.environmentContentCurator = environmentContentCurator;
+        this.envRemoved = new ArrayList<>();
+    }
+
+    public EntitlementEnvironmentFilter setPreExistingEnvironments(List<String> preExistingEnvs) {
+        this.preExistingEnvs = preExistingEnvs;
+        return this;
+    }
+
+    public EntitlementEnvironmentFilter setUpdatedEnvironment(List<String> updatedEnvs) {
+        this.updatedEnvs = updatedEnvs;
+        return this;
+    }
+
+    public EntitlementEnvironmentFilter setConsumerToBeUpdated(List<String> consumerIds) {
+        this.consumerIds = consumerIds;
+        return this;
+    }
+
+    /**
+     * Returns the list of environment IDs which got added, removed or re-prioritized
+     * from consumers current existing environments.
+     *
+     * @return
+     *  Returns the unique environmentIds which got added, removed or re-prioritized
+     */
+    private Set<String> getEnvironmentChangeList() {
+        Map<String, Integer> order = new HashMap<>();
+        Set<String> environmentChangeList = new HashSet<>();
+        // Envs Added
+        environmentChangeList.addAll(CollectionUtils.subtract(this.updatedEnvs, this.preExistingEnvs));
+
+        // Env removed are kept in separate list for special use cases
+        envRemoved.addAll(CollectionUtils.subtract(this.preExistingEnvs, this.updatedEnvs));
+        environmentChangeList.addAll(CollectionUtils.subtract(this.preExistingEnvs, this.updatedEnvs));
+
+        // Env Re-ordered
+        for (int counter = 0; counter < this.preExistingEnvs.size(); counter++) {
+            order.put(this.preExistingEnvs.get(counter), counter);
+        }
+
+        for (int counter = 0; counter < this.updatedEnvs.size(); counter++) {
+            String env = this.updatedEnvs.get(counter);
+
+            if (order.containsKey(env)) {
+                if (order.get(env) != counter) {
+                    environmentChangeList.add(env);
+                }
+            }
+        }
```

Comment:
This implementation has the same problem as the previous one; it still will report some items as re-ordered, even though they were not re-ordered (compared to each other): https://gist.github.com/nikosmoum/cdec89f3607ffdf1774a6a671c0d92b3 (the 2 example unit tests here should pass, but they don't)

This is a place that would benefit from some unit testing, but being a private method we can't. I would suggest extracting out this as a separate public static method in a new ListUtil class, that will be unit testable

Reply:
```
List preexisting = List.of("A", "B");
List updated = List.of("A", "C", "B");
```

In above scenario, the Priority of environment B is reduced. Which means that environment B got re-prioritized. Shouldn't we include in the environment change list ?

Reply:
Added the new reordering logic in CP Utils class.